docs/design-tech/sync.md

```diff
@@ -1,89 +1,118 @@
+* Path Conflict Invariant: no two files may have the same name and parent. An example challenge is when a user creates a new file on each of two clients with the same name and parent, then syncs them both.
+* Cycle Invariant: every non-root file must not have itself as an ancestor.  An example challenge is when on one client the user moves folder A into folder B and on the other client moves folder B into folder A, then syncs them both.
+* Orphan Invariant: every non-root file must have its parent in the file tree. To keep client storage from growing unboundedly, clients eventually prune deleted files (remove all mentions of them from durable storage). If the design does not pay careful attention to the disctinctions between explictly deleted, implicitly deleted, and pruned files, a client may prune a file without pruning one of its children.
 
-The local stores of state that are relevant to syncing are:
+_Concern: does the server check that a user's root folder is named after the username?_
 
-+ `FileMetadataRepo` -- list of all the file metadata objects known locally
-+ `DocumentRepo` -- store of every document known locally
-+ `LocalChangesRepo` -- a description of all the changes that have been made locally
-+ `LastSyncedRepo` -- a simple `get/set` value that the sync algorithm uses to help the server filter out events that
-  are already known locally
+### Privacy
+Files, file names, and encryption keys never leave the client unencrypted. In order to support sharing, we give every document a symmetric key (used to encrypt its contents) and every folder a symmetric key (used to encrypt the keys of child folders and documents). This allows users to recursively share folder contents performantly and consistently. It also creates an encryption chain for each document: the document is encrypted with the document's key, which is encrypted with its parent folder's key, which is encrypted with it's parent folders key, all the way up to the root folder whose key is encrypted with the user's account key which is stored on the user's devices and trasferred directly between them as a string or QR code.
 
-As discussed in the [data model](data_model.md), every time a file's property changes, the server increments its
-`metadata_version`. Ordering the `file_metadata` by this property is our primary means of filtering out what events a
-client knows about and what new events it has to sync.
+This encryption chain, and its interaction with the invariants, creates programming challenges. Our general pattern is to decrypt keys, names, and files as we read them from disk or as responses from the server so that we can write logic without worrying about encryption. However, most invariant violations (e.g. orphans and cycles) will cause us to fail to be able to decrypt files, which in turn makes it difficult to detect and appropriately resolve invariant violations.
```

Comment:
```suggestion
This encryption chain, and its interaction with the invariants, creates programming challenges. Our general pattern is to decrypt keys, names, and files as we read them from disk or receive them as responses from the server so that we can write logic without worrying about encryption. However, most invariant violations (e.g. orphans and cycles) will cause us to fail to be able to decrypt files, which in turn makes it difficult to detect and appropriately resolve invariant violations.
```